staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go

```diff
@@ -0,0 +1,544 @@
+	defer ctcTearDown()
+
+	// read only object to read at a different version than stored when we need to force conversion
+	marker, err := ctc.versionedClient("marker", "v1beta2").Create(NewMultiVersionInstance("marker", "marker", "v1beta2"), metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, test := range tests {
+		t.Run(test.group, func(t *testing.T) {
+			tearDown, webhookClientConfig, webhookWaitReady, err := fixtures.StartConversionWebhookServerWithWaitReady(test.params.handler)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer tearDown()
+
+			ctc.setConversionWebhook(t, allVersions, webhookClientConfig)
+			defer ctc.removeConversionWebhook(t)
+
+			err = webhookWaitReady(30*time.Second, func() error {
+				// the marker's storage version is v1beta2, so a v1beta1 read always triggers conversion
+				ctc.versionedClient(marker.GetNamespace(), "v1beta1").Get(marker.GetName(), metav1.GetOptions{})
+				return nil
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			for i, checkFn := range test.checks {
+				name := fmt.Sprintf("check-%d", i)
+				t.Run(name, func(t *testing.T) {
+					ctc.namespace = fmt.Sprintf("webhook-conversion-%s-%s", test.group, name)
+					checkFn(t, ctc)
+				})
+			}
+		})
+	}
+}
+
+func validateStorageVersion(t *testing.T, ctc *conversionTestContext) {
+	ns := ctc.namespace
+	defer ctc.setStorageVersion(t, "v1beta2")
+
+	for _, version := range []string{"v1beta1", "v1beta2"} {
+		t.Run(version, func(t *testing.T) {
+			name := "storageversion-" + version
+			client := ctc.versionedClient(ns, version)
+			obj, err := client.Create(NewMultiVersionInstance(ns, name, version), metav1.CreateOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+			ctc.setStorageVersion(t, "v1beta2")
+			ctc.waitForStorageVersion(t, "v1beta2", client, obj)
+
+			obj, err = client.Get(obj.GetName(), metav1.GetOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ctc.setStorageVersion(t, "v1beta1")
+			ctc.waitForStorageVersion(t, "v1beta1", client, obj)
+		})
+	}
+}
+
+func validateMixedStorageVersions(t *testing.T, ctc *conversionTestContext) {
+	ns := ctc.namespace
+	defer ctc.setStorageVersion(t, "v1beta2")
+
+	v1client := ctc.versionedClient(ns, "v1beta1")
+	v2client := ctc.versionedClient(ns, "v1beta2")
+	clients := map[string]dynamic.ResourceInterface{"v1beta1": v1client, "v1beta2": v2client}
+	versions := []string{"v1beta1", "v1beta2"}
+	for _, version := range versions {
+		name := "stored-at-" + version
+		obj, err := v1client.Create(NewMultiVersionInstance(ns, name, "v1beta1"), metav1.CreateOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+		ctc.setStorageVersion(t, version)
+		ctc.waitForStorageVersion(t, version, v1client, obj)
+	}
+
+	for clientVersion, client := range clients {
+		t.Run(clientVersion, func(t *testing.T) {
+			objs := []*unstructured.Unstructured{}
+			for _, version := range versions {
+				obj, err := client.Get("stored-at-"+version, metav1.GetOptions{})
```

Comment:
this tests "persist some data, change the storage version, and ensure the previously persisted data is readable"

Reply:
Added a function level comment.